hal/src/nRF52840/ble_hal.cpp

```diff
@@ -15,739 +15,3147 @@
+        len = 0;
     }
-    // Set data length size
-    ret = nrf_ble_gatt_data_length_set(&g_gatt, BLE_CONN_HANDLE_INVALID, DEFAULT_DATA_LENGTH);
-    if (ret != NRF_SUCCESS) {
-        LOG(ERROR, "nrf_ble_gatt_data_length_set() failed: %u", (unsigned)ret);
-        return halError(ret);
+    scanRespDataLen_ = len;
+    ret = configure(nullptr);
+    CHECK_RETURN(ret, nullptr);
+    return resume();
+}
+
+size_t BleObject::Broadcaster::getScanResponseData(uint8_t* buf, size_t len) const {
+    if (buf == nullptr) {
+        return 0;
     }
-    // Set MTU size
-    ret = nrf_ble_gatt_att_mtu_periph_set(&g_gatt, DEFAULT_MTU);
-    if (ret != NRF_SUCCESS) {
-        LOG(ERROR, "nrf_ble_gatt_att_mtu_periph_set() failed: %u", (unsigned)ret);
-        return halError(ret);
+    len = MIN(len, scanRespDataLen_);
+    memcpy(buf, scanRespData_, len);
+    return len;
+}
+
+int BleObject::Broadcaster::setTxPower(int8_t val) {
+    val = roundTxPower(val);
+    int ret = sd_ble_gap_tx_power_set(BLE_GAP_TX_POWER_ROLE_ADV, advHandle_, val);
+    NRF_CHECK_RETURN(ret, "sd_ble_gap_tx_power_set");
+    txPower_ = val;
+    return SYSTEM_ERROR_NONE;
+}
+
+int8_t BleObject::Broadcaster::getTxPower(void) const {
+    return txPower_;
+}
+
+int BleObject::Broadcaster::startAdvertising(void) {
+    int ret;
+    if (isAdvertising_) {
+        // Stop advertising first.
+        ret = sd_ble_gap_adv_stop(advHandle_);
+        NRF_CHECK_RETURN(ret, "sd_ble_gap_adv_stop");
+        isAdvertising_ = false;
     }
-    return 0;
+    ret = sd_ble_gap_adv_start(advHandle_, BLE_CONN_CFG_TAG);
+    NRF_CHECK_RETURN(ret, "sd_ble_gap_adv_start");
+    isAdvertising_ = true;
+    return SYSTEM_ERROR_NONE;
 }
 
-int initGap(const char* deviceName) {
-    ble_gap_conn_sec_mode_t secMode = {};
-    // Require no protection
-    BLE_GAP_CONN_SEC_MODE_SET_OPEN(&secMode);
-    // Set device name
-    uint32_t ret = sd_ble_gap_device_name_set(&secMode, (const uint8_t*)deviceName, strlen(deviceName));
-    if (ret != NRF_SUCCESS) {
-        LOG(ERROR, "sd_ble_gap_device_name_set() failed: %u", (unsigned)ret);
-        return halError(ret);
-    }
-    // Set PPCP (Peripheral Preferred Connection Parameters)
-    ble_gap_conn_params_t connParam = {};
-    connParam.min_conn_interval = MIN_CONN_INTERVAL;
-    connParam.max_conn_interval = MAX_CONN_INTERVAL;
-    connParam.conn_sup_timeout = CONN_SUP_TIMEOUT;
-    connParam.slave_latency = SLAVE_LATENCY;
-    ret = sd_ble_gap_ppcp_set(&connParam);
-    if (ret != NRF_SUCCESS) {
-        LOG(ERROR, "sd_ble_gap_ppcp_set() failed: %u", (unsigned)ret);
-        return halError(ret);
+int BleObject::Broadcaster::stopAdvertising(void) {
+    if (isAdvertising_) {
+        int ret = sd_ble_gap_adv_stop(advHandle_);
+        NRF_CHECK_RETURN(ret, "sd_ble_gap_adv_stop");
+        isAdvertising_ = false;
     }
-    return 0;
+    return SYSTEM_ERROR_NONE;
 }
 
-} // ::
+int BleObject::Broadcaster::suspend(void) {
+    if (isAdvertising_) {
```

Comment:
Should `isAdvertising_` also go to `false`? If so or doesn't matter, consider simply calling `stopAdvertising()`.